ducktape/tests/runner_client.py

```diff
@@ -102,82 +105,89 @@ def run(self):
-            for node_spec in min_cluster_spec:
-                if not os_to_num_nodes.get(node_spec.operating_system):
-                    os_to_num_nodes[node_spec.operating_system] = 1
-                else:
-                    os_to_num_nodes[node_spec.operating_system] = os_to_num_nodes[node_spec.operating_system] + 1
-            for (operating_system, node_count) in iteritems(os_to_num_nodes):
-                num_avail = len(list(self.cluster.all().nodes.elements(operating_system=operating_system)))
-                if node_count > num_avail:
-                    raise RuntimeError(
-                        "There are not enough nodes available in the cluster to run this test. "
-                        "Cluster size for %s: %d, Need at least: %d. Services currently registered: %s" %
-                        (operating_system, num_avail, node_count, self.test_context.services))
-
-            # Run the test unit
-            start_time = time.time()
-            self.setup_test()
-
-            data = self.run_test()
-
-            test_status = PASS
-            self.log(logging.INFO, "PASS")
-
-        except BaseException as e:
-            # mark the test as failed before doing anything else
-            test_status = FAIL
-            err_trace = self._exc_msg(e)
-            summary += err_trace
-            self.log(logging.INFO, "FAIL: " + err_trace)
-
-        finally:
-            self.teardown_test(teardown_services=not self.session_context.no_teardown, test_status=test_status)
-
-            stop_time = time.time()
-
-            if hasattr(self, "services"):
-                service_errors = self.test_context.services.errors()
-                if service_errors:
-                    summary += "\n\n" + service_errors
-
-            test_status, summary = self._check_cluster_utilization(test_status, summary)
-
-            result = TestResult(
-                self.test_context,
-                self.test_index,
-                self.session_context,
-                test_status,
-                summary,
-                data,
-                start_time,
-                stop_time)
-
-            self.log(logging.INFO, "Summary: %s" % str(result.summary))
-            self.log(logging.INFO, "Data: %s" % str(result.data))
-
-            result.report()
-            # Tell the server we are finished
-            self._do_safely(lambda: self.send(self.message.finished(result=result)),
-                            "Problem sending FINISHED message for " + str(self.test_metadata) + ":\n")
-            # Release test_context resources only after creating the result and finishing logging activity
-            # The Sender object uses the same logger, so we postpone closing until after the finished message is sent
-            self.test_context.close()
-            self.test_context = None
-            self.test = None
+        num_runs = self.deflake_num
+        while test_status == FAIL and num_runs > 0:
```

Comment:
This is a little bit difficult to read, with num_runs counting backwards and num_runs == self.deflake_num - 1 later on, when we can simply do num_runs < deflake_num and then num_runs == 1 below.